src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8376,283 @@ private StatementSyntax ParseLocalDeclarationStatement()
+        {
+            var identifier = ParseIdentifierToken();
+            var declarator = _syntaxFactory.VariableDeclarator(identifier, null, null);
+            return _syntaxFactory.VariableDeclaration(
+                                    type: null,
+                                    variables: SyntaxFactory.SeparatedList(declarator),
+                                    deconstruction: null);
+        }
+
+        /// <summary>
+        /// Parses an individual type and identifier in a deconstruction declaration. For instance, in `(type id, type id) = ...`
+        /// Never returns null.
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionTypeIdPart()
+        {
+            var type = ParseType(parentIsParameter: false);
+            var identifier = ParseIdentifierToken();
+
+            var declarator = _syntaxFactory.VariableDeclarator(identifier, null, null);
+            return _syntaxFactory.VariableDeclaration(
+                                    type: type,
+                                    variables: SyntaxFactory.SeparatedList(declarator),
+                                    deconstruction: null);
+        }
+
+        /// <summary>
+        /// Check ahead for a deconstruction declaration. This requires at least one good-looking variable and the presence of an equals sign.
+        /// Doesn't move the cursor.
+        /// PROTOTYPE(tuples) Can this be done without allocations?
+        /// </summary>
+        private bool IsPossibleDeconstructionDeclaration()
+        {
+            if ((this.CurrentToken.IsVar() && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken) ||
+                this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                var resetPoint = this.GetResetPoint();
+                try
+                {
+                    // We don't need to parse the expression following the equals token
+                    var variables = ParseDeconstructionDeclaration(withEquals: false);
+                    var equalsToken = this.EatToken(SyntaxKind.EqualsToken);
+
+                    // We just need the equals token and one other confirmation that this is a deconstruction syntax
+                    return DeconstructionVariableLooksGood(variables, withEquals: false) && !equalsToken.IsMissing;
+                }
+                finally
+                {
+                    this.Reset(ref resetPoint);
+                    this.Release(ref resetPoint);
+                }
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if one distinct clue is found that this is intended as deconstruction variables, and the equals sign is as expected.
+        /// </summary>
+        private static bool DeconstructionVariableLooksGood(VariableDeclarationSyntax node, bool withEquals = true)
```

Comment:
`ParseDeconstructionDeclaration` is recursive (`ParseDeconstructionList` calls it and gets called by it), so we'd also have to carry around one more flag to remember to do the LooksGood check. I'm not sure it would simplify the code.